### ISSUE TITLE
Inconsistent behavior between pycoco and keras_cv.metrics.COCOMeanAveragePrecision

### ISSUE THREAD

**Issue by tanzhenyu** [keras-team/keras-cv#915]
During the development of FasterRCNN, I have noticed reasonable loss and reasonable prediction outcome, but always get low mAP metrics (~0.18). The evaluation loop takes ~13 minutes.
So I instead hack a COCO wrapper (thanks to Model Garden) and directly call pycoco, this time it gets me 0.44 mAP which is more reasonable. The evaluation loop takes ~8 minutes.
I wonder how we can investigate those inconsistencies.
I have put my reproducible code in a separate branch, this is a standalone script that can run end-to-end:
https://github.com/keras-team/keras-cv/blob/map/examples/models/object_detection/faster_rcnn/eval.py

I have also tried to configure iou_thresholds, recall_thresholds to try to match the result from pycoco, but with no luck.

You can also copy the script to colab if you want to run the `visualize_image` section along with mAP computation.

**Comment by LukeWood** [comment]
> During the development of FasterRCNN, I have noticed reasonable loss and reasonable prediction outcome, but always get low mAP metrics (~0.18). The evaluation loop takes ~13 minutes. So I instead hack a COCO wrapper (thanks to Model Garden) and directly call pycoco, this time it gets me 0.44 mAP which is more reasonable. The evaluation loop takes ~8 minutes. I wonder how we can investigate those inconsistencies. I have put my reproducible code in a separate branch, this is a standalone script that can run end-to-end: https://github.com/keras-team/keras-cv/blob/map/examples/models/object_detection/faster_rcnn/eval.py
> 
> I have also tried to configure iou_thresholds, recall_thresholds to try to match the result from pycoco, but with no luck.
> 
> You can also copy the script to colab if you want to run the `visualize_image` section along with mAP computation.

Thanks for the error report!  I'll take this on and try to figure out whats going on here.

**Comment by bhack** [comment]
> The evaluation loop takes ~8 minutes.

@tanzhenyu Other then precision that could be fixed/covered by the right test we still have some perf. inefficiency as we cannot XLA compile these things. See https://github.com/keras-team/keras-cv/issues/895#issuecomment-1273430678

**Comment by tanzhenyu** [comment]
> n loop takes ~8 minutes.

Those metrics that require lots of dynamic processing are usually something we cannot make it happen in XLA anyway, so right now this only works on CPU, we don't have plans to support it in GPU/TPU

**Comment by tanzhenyu** [comment]
> > The evaluation loop takes ~8 minutes.
> 
> @tanzhenyu Other then precision that could be fixed/covered with the right test coverage we still have some perf. inefficiency as we cannot XLA compile these things. See [#895 (comment)](https://github.com/keras-team/keras-cv/issues/895#issuecomment-1273430678)

Those metrics that require lots of dynamic processing are usually something we cannot make it happen in XLA anyway, so right now this only works on CPU, we don't have plans to support it in GPU/TPU

**Comment by bhack** [comment]
> Those metrics that require lots of dynamic processing are usually something we cannot make it happen in XLA anyway

IMHO It is not clear as initially we had a "simple" issue related to the missing coverage of `tf.bincount` in the bridge so I have contributed the kernel in Tensorflow.

Currently we are investigating on https://github.com/tensorflow/tensorflow/issues/58051

**Comment by LukeWood** [comment]
> During the development of FasterRCNN, I have noticed reasonable loss and reasonable prediction outcome, but always get low mAP metrics (~0.18). The evaluation loop takes ~13 minutes. So I instead hack a COCO wrapper (thanks to Model Garden) and directly call pycoco, this time it gets me 0.44 mAP which is more reasonable. The evaluation loop takes ~8 minutes. I wonder how we can investigate those inconsistencies. I have put my reproducible code in a separate branch, this is a standalone script that can run end-to-end: https://github.com/keras-team/keras-cv/blob/map/examples/models/object_detection/faster_rcnn/eval.py
> 
> I have also tried to configure iou_thresholds, recall_thresholds to try to match the result from pycoco, but with no luck.
> 
> You can also copy the script to colab if you want to run the `visualize_image` section along with mAP computation.

Some notes:

- re: https://github.com/keras-team/keras-cv/blob/map/examples/models/object_detection/faster_rcnn/eval.py#L448 this is just part of the definition of MaP.  Check the technical report I wrote on the metrics, I think it is a good explanation in the metric definition section: https://arxiv.org/abs/2207.12120. the metric is NOT a substitute for precision despite the name

- re the performance: you will want to pad your tensors in the model, and run the update in graph mode.  Right now I think its all being run eagerly.

- if you print out the GT classes and CLS classes between your two methods you get different results:  
```
gt_classes tf.Tensor(
[[[ 8.]
  [15.]
  [15.]]], shape=(1, 3, 1), dtype=float32)
cls_classes tf.Tensor([[ 8. 15. 15.  8.  0.  0.  0.  0.  0.  0.]], shape=(1, 10), dtype=float32)

In KerasCV step
gt_classes tf.Tensor(
[[[ 8.]
  [15.]
  [15.]]], shape=(1, 3, 1), dtype=float32)
cls_pred tf.Tensor(
[[[ 8.]
  [15.]
  [15.]
  [ 8.]]], shape=(1, 4, 1), dtype=float32)
```

**Comment by tanzhenyu** [comment]
> * https://arxiv.org/abs/2207.12120

So pycoco is known to be able to deal with padded 0 values, so passing
`cls_classes tf.Tensor([[ 8. 15. 15.  8.  0.  0.  0.  0.  0.  0.]], shape=(1, 10), dtype=float32)`
to it seems give me exactly result compared to passing
`cls_classes tf.Tensor([[ 8. 15. 15.  8.]], shape=(1, 4), dtype=float32)`

I wasn't so sure if keras_cv.COCO can support it (the documentation says it should though), so I only passed valid values instead to make sure not messing up with its result

**Comment by LukeWood** [comment]
> > * https://arxiv.org/abs/2207.12120
> 
> So pycoco is known to be able to deal with padded 0 values, so passing `cls_classes tf.Tensor([[ 8. 15. 15. 8. 0. 0. 0. 0. 0. 0.]], shape=(1, 10), dtype=float32)` to it seems give me exactly result compared to passing `cls_classes tf.Tensor([[ 8. 15. 15. 8.]], shape=(1, 4), dtype=float32)`
> 
> I wasn't so sure if keras_cv.COCO can support it (the documentation says it should though), so I only passed valid values instead to make sure not messing up with its result

Thanks for the explanation.  It handles -1 padded, but not zero.  Zero would in fact mess the result up. 

 I’ll keep digging on the discrepancy…

**Comment by bhack** [comment]
https://github.com/keras-team/keras-cv/blob/e9390abbf12f2a906e79fb773f69fc0e76821340/keras_cv/metrics/coco/mean_average_precision.py#L68-L73

**Comment by tanzhenyu** [comment]
> https://github.com/keras-team/keras-cv/blob/e9390abbf12f2a906e79fb773f69fc0e76821340/keras_cv/metrics/coco/mean_average_precision.py#L68-L73

Oh, maybe we should add a note or warning here saying accepting ragged tensor will trigger retracing, and prefer users to pad with -1?

**Comment by bhack** [comment]
P:s. ragged tensors are also not supported by the standard TF XLA bridge

**Comment by LukeWood** [comment]
> > https://github.com/keras-team/keras-cv/blob/e9390abbf12f2a906e79fb773f69fc0e76821340/keras_cv/metrics/coco/mean_average_precision.py#L68-L73
> 
> Oh, maybe we should add a note or warning here saying accepting ragged tensor will trigger retracing, and prefer users to pad with -1?

Good point.  I don’t * think * that retracing is always the case with raggeds.  Perhaps we need to put relax_shapes=True in the tf.function decorator on the COCO metrics’ update state methods.  I’ll give that a try.

**Comment by LukeWood** [comment]
Ok, there is definitely a but in the KerasCV COCO metrics.  Its interesting, but in the start of their runs the numbers are very close to the `pycocotools` numbers (i.e. when you've only evaluated 10-20 batches), but over time the numbers tend to decrease instead of remain consistent! 

So `model.evaluate(eval_ds.take(10))` consistently aligns with the pycocotools result (with some small errors), but `model.evaluate(eval_ds)` consistently scores 50% lower than pycocotools. 

quite interesting, I'll keep digging.

**Comment by bhack** [comment]
> Ok, there is definitely a but in the KerasCV COCO metrics. Its interesting, but in the start of their runs the numbers are very close to the `pycocotools` numbers (i.e. when you've only evaluated 10-20 batches), but over time the numbers tend to decrease instead of remain consistent!
> 
> So `model.evaluate(eval_ds.take(10))` consistently aligns with the pycocotools result (with some small errors), but `model.evaluate(eval_ds)` consistently scores 50% lower than pycocotools.
> 
> quite interesting, I'll keep digging.

Do you think it is possible that we are accumulating [known small numerical error](https://github.com/keras-team/keras-cv/issues/149#issuecomment-1104532810) over the whole eval ds?

**Comment by tanzhenyu** [comment]
> Ok, there is definitely a but in the KerasCV COCO metrics. Its interesting, but in the start of their runs the numbers are very close to the `pycocotools` numbers (i.e. when you've only evaluated 10-20 batches), but over time the numbers tend to decrease instead of remain consistent!
> 
> So `model.evaluate(eval_ds.take(10))` consistently aligns with the pycocotools result (with some small errors), but `model.evaluate(eval_ds)` consistently scores 50% lower than pycocotools.
> 
> quite interesting, I'll keep digging.

I had the same observation

**Comment by LukeWood** [comment]
> > Ok, there is definitely a but in the KerasCV COCO metrics. Its interesting, but in the start of their runs the numbers are very close to the `pycocotools` numbers (i.e. when you've only evaluated 10-20 batches), but over time the numbers tend to decrease instead of remain consistent!
> > So `model.evaluate(eval_ds.take(10))` consistently aligns with the pycocotools result (with some small errors), but `model.evaluate(eval_ds)` consistently scores 50% lower than pycocotools.
> > quite interesting, I'll keep digging.
> 
> Do you think it is possible that we are accumulating [known small numerical error](https://github.com/keras-team/keras-cv/issues/149#issuecomment-1104532810) over the whole eval ds?

I never implemented https://github.com/keras-team/keras-cv/issues/149, so that wouldn't be the issue.   https://github.com/keras-team/keras-cv/issues/149 would be a deviation from the original to improve our own numerical accuracy even over the pycocotools API - but I never implemented it.
